### PR TITLE
sway: refactor with a wrapper

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

When executing `sway`, if you start `waybar`, or `wofi` , Icon themes and/or Gtk Themes are not loaded.

It can be fixed by bulindg Sway with a `wrapGAppsHook` in `nativeBuildInputs`. @primeos proposed on issue #67704 to default on an unwrapped version. which is made is this PR with `programs.sway.package` option defaulting to an `unwrapped` version of Sway.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @primeos

### CODE REVIEW

File: nixos/modules/programs/sway.nix
Comment (symphorien) [inline nixos/modules/programs/sway.nix]:
```suggestion
      defaultText = "pkgs.sway-unwrapped";
```

Comment (Elyhaka) [reply nixos/modules/programs/sway.nix]: Fixed it, thank you.

File: nixos/modules/programs/sway.nix
Comment (primeos) [inline nixos/modules/programs/sway.nix]: Not sure if we need a package option (normally we try to avoid it in cases like this where there are no alternative packages (like e.g. sway-gaps)), but I guess it could become useful if the wrapper has many "use flags" (like `withGAppsWrapper`) and we provide an example.

File: pkgs/top-level/all-packages.nix
Comment (primeos) [inline pkgs/top-level/all-packages.nix]: I think what I meant was that `sway` would take `sway-unwrapped` as input/dependency and add a wrapper around `bin/sway`. This would have the advantage that one can customize the wrapper without having to rebuild Sway (it doesn't take very long, but I think it would be nice to avoid this).

I'm not totally sure but I think it would be best if we also move the wrapping for `swaybg` here and the functionality from the NixOS module as well (-> we would have only one wrapper instead of 2-3 and non-NixOS users could also use part of the functionality of the NixOS module, like the D-BUS part).

**Review by primeos — CHANGES_REQUESTED** [review]
Sorry for the delay, I'm a bit short on time atm.

This differs from what I had in mind, but I'm also open for suggestions if you or someone else has a better idea (I haven't thought that much about this so far). If you just want the functionality I'm also fine if this only changes the `sway` package by e.g. introducing a `wrapGApps` "use flag" (true/false).

Feel free to ask any questions, my review might be too short to properly explain what I had in mind... :o

**Comment by Elyhaka** [comment]
Hello @primeos, thank you for your extensive review ! I've re-read what you said on #67704 and with the information you given I've changed the way I implemented it :

 - We now have a wrapper separated from sway itself.
 - I've moved the wrapper from the `sway` module (`programs.sway`) to the wrapper itself.
    - It is enabled by default, and it has a toggle named `withBaseWrapper`.
 - I've added a toggle named  `withGAppsWrapper`, disabled by default.

I did not understood what you meant with the `swaybg` part on your last review. Could you please clarify this point ? :)

**Comment by Synthetica9** [comment]
Maybe rename the `withGAppsWrapper` to something like `withGTKWrapper`? GApps is already widely used for Google Apps on the Android platform, I don't think I've ever heard it used in reference to GTK.

**Comment by Elyhaka** [comment]
@Synthetica9 You are right, It wasn't really explicit. Fixed it.

**Comment by Elyhaka** [comment]
@primeos I have just found your PR about adding Sway to dislpay managers :

```
    # To make a Sway session available if a display manager like SDDM is enabled:
    services.xserver.displayManager.extraSessionFilePackages = [ swayJoined ];
```

I believe it will be impacted by this PR (if merged). Do you want to merge yours first then let me rebase mine or the other way around ?

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/wrapper.nix]: I think this belongs into `nativeBuildInputs`, then we can also drop `postBuild` (IIRC).

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/wrapper.nix]: This shouldn't be necessary. Is it for other reasons (e.g. to easily get Sway's build dependencies, though we could get them though `sway-unwrapped`)?

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/wrapper.nix]: This package uses 2x `symlinkJoin` and we don't need to join anything here. It would be great if we would only need to use it once to add the `baseWrapper`.

**Review by primeos — CHANGES_REQUESTED** [review]
Awesome, looks way better now :)

Regarding the other PR: I'll probably merge my PR first as it's very trivial and already good to go (I assume). Feel free to rebase if you want.

**Comment by primeos** [comment]
cc @gnidorah @colemickens (FYI, forgot that `@GrahamcOfBorg` doesn't notify module maintainers, feel free to add feedback)

File: nixos/modules/programs/sway.nix
Comment (Elyhaka) [reply nixos/modules/programs/sway.nix]: I've implemented the second proposition here :).

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: `pname` did not work with `symlinkJoin`. Used the second option.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: If I drop the `postBuild` the GTK Wrapper seems to miss some variables (not really sure why) and it revert to the previous behavior.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Like for the `postBuild` hook, if I drop the `sway-unwrapped.buildInputs` the GTK Wrapper seems to miss some variables and `gsettings` seems to stop working again.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Not crashing, just displaying a gray background, I've moved it.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Did you move `wrapGAppsHook` to `nativeBuildInputs`?

**Comment by Elyhaka** [comment]
Thank you again @primeos for your time and review, I've made the fixes you asked, important changes made in this iteration :

 - We now have a `wrapperFeatures` that can be used like you proposed (it's way cleaner).
 - I've moved `swaybg` to the wrapper.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Indeed, even moved it was not triggered. I have the feeling that `symlinkJoin` does not trigger automatically hooks from `nativeBuildInputs`. Am I correct ?

Comment (primeos) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Haven't looked into it but that assumption sounds reasonable.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: @primeos I've just checked https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/wrap-gapps-hook.sh and it seems it uses the variable present at build time to reexport them in a wrapper.
I might be compltely wrong, but since I use a `symlinkJoin`, could it be that the environment exposed while creating the wrapper is not the same, thus, it's not wrapping correctly `sway-unwrapped` ?

**Comment by Elyhaka** [comment]
Rebased branch after merging of #75363

**Comment by primeos** [comment]
@Elyhaka thanks for your awesome work, I finally got around to properly review this (sorry for the long delay :o) and added a few minor improvements. Are you ok with these changes? If so I would run a final test, squash my commit into yours and it should be ready to be merged.

File: nixos/modules/programs/sway.nix
Comment (primeos) [inline nixos/modules/programs/sway.nix]: Oh dear, seems like I've messed up the indentation here... :o (probably tabs vs. spaces due to the default Vim configuration of my test VM)

File: nixos/modules/programs/sway.nix
Comment (primeos) [inline nixos/modules/programs/sway.nix]: Thought this could be useful to avoid mistakes and redundancy. Also e.g. `example = "false";` didn't need any quoting.

File: nixos/modules/programs/sway.nix
Comment (primeos) [inline nixos/modules/programs/sway.nix]: Indentation, again...

File: nixos/modules/programs/sway.nix
Comment (primeos) [inline nixos/modules/programs/sway.nix]: Indentation

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/wrapper.nix]: This is pretty dirty tbh... :o The result should be fine, but the code is probably difficult to understand... I couldn't find a nicer way to write this (without adding more abstractions).

**Review by primeos — CHANGES_REQUESTED** [review]
~~Just a quick review for myself to not forget about the wrong indentations I've introduced...~~

**Update:** Fixed the indentations.

**TODOs:**
- [x] ACK from @Elyhaka
- [x] Squash the commits
- [x] Final test

File: nixos/modules/programs/sway.nix
Comment (Elyhaka) [reply nixos/modules/programs/sway.nix]: This is really clean, I didn't know it could be written this way, thanks for showing it :)

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Elyhaka) [reply pkgs/applications/window-managers/sway/wrapper.nix]: I see, I had to read it twice to understand the intent (but I'm not yet comfortable with all Nix subtleties), maybe we could let a comment here ?

**Comment by Elyhaka** [comment]
This looks good to me ! Thank you for taking the time to do the final changes :smile:

**Comment by Elyhaka** [comment]
@primeos : I've made some tests, and it seems we have a regression on gsettings. I'm investigating the cause.

**Comment by Elyhaka** [comment]
@primeos : It works with the last commit :+1:

**Comment by primeos** [comment]
@Elyhaka thanks, I totally missed that regression :o I verified that the wrapper was adding arguments but didn't realise that some where missing. I've had a look at `pkgs/build-support/setup-hooks/wrap-gapps-hook.sh` and noticed that `wrapGAppsHook` must be executed to initialize the variables. It should work as expected now. I've also updated `buildInputs` as it turns out that only `gdk-pixbuf` should be required.

The reason for the `dontWrapGApps=true` hack is that I would like to avoid wrapping all binaries and wrapping sway twice:
```
[root@nixos-unstable:/var/nixpkgs]# ls -al /nix/store/k4cm6pvbc37iwz9ac6vb38kvzpwipczc-sway-1.2/bin/
total 44
dr-xr-xr-x 2 root root 4096 Jan  1  1970 .
dr-xr-xr-x 5 root root 4096 Jan  1  1970 ..
-r-xr-xr-x 1 root root 1047 Jan  1  1970 sway
-r-xr-xr-x 1 root root  959 Jan  1  1970 swaybar
lrwxrwxrwx 1 root root   74 Jan  1  1970 .swaybar-wrapped -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaybar
-r-xr-xr-x 1 root root  959 Jan  1  1970 swaymsg
lrwxrwxrwx 1 root root   74 Jan  1  1970 .swaymsg-wrapped -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaymsg
-r-xr-xr-x 1 root root  959 Jan  1  1970 swaynag
lrwxrwxrwx 1 root root   74 Jan  1  1970 .swaynag-wrapped -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaynag
lrwxrwxrwx 1 root root   71 Jan  1  1970 .sway-wrapped -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/sway
-r-xr-xr-x 1 root root  956 Jan  1  1970 .sway-wrapped_
```

With this hack only sway will be wrapped and also only once (if I think about it it should also be possible to get rid of `.sway-wrapped` by using something like `makeWrapper $oldBinary $wrapper $args` but that could be optimized later):
```
[root@nixos-unstable:/var/nixpkgs]# ls -al /nix/store/0g4gkad9ikim9hf94sgzn9g3dgnganbv-sway-1.2/bin/
total 28
dr-xr-xr-x 2 root root 4096 Jan  1  1970 .
dr-xr-xr-x 5 root root 4096 Jan  1  1970 ..
-r-xr-xr-x 1 root root 1046 Jan  1  1970 sway
lrwxrwxrwx 1 root root   74 Jan  1  1970 swaybar -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaybar
lrwxrwxrwx 1 root root   74 Jan  1  1970 swaymsg -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaymsg
lrwxrwxrwx 1 root root   74 Jan  1  1970 swaynag -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/swaynag
lrwxrwxrwx 1 root root   71 Jan  1  1970 .sway-wrapped -> /nix/store/86gzfk3pj8bg4clw2var203id2br8wr2-sway-unwrapped-1.2/bin/sway
```

**Comment by Elyhaka** [comment]
@primeos Wow, really good catch ! Thank you very much for the explanation. It is way cleaner with this approach. Everything looks good to me !

**Review by primeos — APPROVED** [review]
Should be ready to merge now :) \o/

Thank you @Elyhaka

**Comment by Elyhaka** [comment]
Thank you too @primeos ! :smiley:

**Pull request merged**
